taichi/rhi/metal/metal_device.mm

```diff
@@ -830,17 +890,197 @@ DeviceCapabilityConfig collect_metal_device_caps(MTLDevice_id mtl_device) {
+             (src[0].stage == PipelineStageType::vertex &&
+              src[1].stage == PipelineStageType::fragment));
+
+  spirv_cross::CompilerMSL::Options options{};
+  options.enable_decoration_binding = true;
+
+  // Compile spirv binaries to MSL source
+  std::string msl_source = "";
+  for (int i = 0; i < 2; i++) {
+    const uint32_t *spv_data = (const uint32_t *)src[i].data;
+
+    RHI_ASSERT((size_t)spv_data % sizeof(uint32_t) == 0);
+    RHI_ASSERT(src[i].size % sizeof(uint32_t) == 0);
+
+    spirv_cross::CompilerMSL compiler(spv_data, src[i].size / sizeof(uint32_t));
+    compiler.set_msl_options(options);
+
+    std::string msl = "";
+    try {
+      msl = compiler.compile();
+    } catch (const spirv_cross::CompilerError &e) {
+      std::array<char, 4096> msgbuf;
+      snprintf(msgbuf.data(), msgbuf.size(), "(spirv-cross compiler) %s: %s",
+               name.c_str(), e.what());
+      RHI_LOG_ERROR(msgbuf.data());
+      return nullptr;
+    }
+
+    const std::string new_entry_point_name =
+        src[i].stage == PipelineStageType::fragment ? frag_function_name
+                                                    : vert_function_name;
+
+    // Fragment and vertex function names must be different.
+    // If spirv-cross has a method to let you set the emitted entry point's
+    // name, that would be nice, but I could not find anything in the docs.
+    // So for now just using std's regex_replace().
+    std::regex entry_point_regex("main0");
+    msl_source +=
+        std::regex_replace(msl, entry_point_regex, new_entry_point_name);
+  }
+
+  // Compile MSL source
+  MTLLibrary_id mtl_library = get_mtl_library(msl_source);
+
+  // Get the MTLFunctions
+  int frag_func_index = 0;
+  int vert_func_index = 0;
+  std::vector<MTLFunction_id> mtl_functions;
+  for (int i = 0; i < 2; i++) {
+    std::string new_entry_point_name;
+    if (src[i].stage == PipelineStageType::fragment) {
+      frag_func_index = i;
+      new_entry_point_name = frag_function_name;
+    } else {
+      vert_func_index = i;
+      new_entry_point_name = vert_function_name;
+    }
+
+    MTLFunction_id mtl_function =
+        get_mtl_function(mtl_library, new_entry_point_name);
+    mtl_functions.push_back(mtl_function);
+  }
+
+  // Create render pipeline
+  static const std::unordered_map<TopologyType, MTLPrimitiveTopologyClass>
+      topo_types = {
+          {TopologyType::Triangles,
+           MTLPrimitiveTopologyClass::MTLPrimitiveTopologyClassTriangle},
+          {TopologyType::Lines,
+           MTLPrimitiveTopologyClass::MTLPrimitiveTopologyClassLine},
+          {TopologyType::Points,
+           MTLPrimitiveTopologyClass::MTLPrimitiveTopologyClassPoint},
+      };
+
+  MTLRenderPipelineDescriptor *rpd = [MTLRenderPipelineDescriptor new];
+  rpd.vertexFunction = mtl_functions[vert_func_index];
+  rpd.fragmentFunction = mtl_functions[frag_func_index];
+  rpd.colorAttachments[0].pixelFormat = format2mtl(swap_chain_image_format_);
```

Comment:
Maybe leave a TODO or FIXME here. Forcing the format to swap chain image format might bite us in the bum later

Reply:
Usually what I do here is to leave this pipeline partially constructed, and uses a map between renderpass and actual pipeline states, and build / cache pipeline states in the `bind_pipeline` command. Because the color attachment format is not known until `begin_renderpass`

Reply:
Don't need to fix it rn tho